ding/policy/ppg.py

```diff
@@ -34,6 +34,589 @@ def create_shuffled_dataloader(data, batch_size):
+
+                - current_lr (:obj:`float`): Current learning rate
+                - total_loss (:obj:`float`): The calculated loss
+                - policy_loss (:obj:`float`): The policy(actor) loss of ppg
+                - value_loss (:obj:`float`): The value(critic) loss of ppg
+                - entropy_loss (:obj:`float`): The entropy loss
+                - auxiliary_loss (:obj:`float`): The auxiliary loss, we use the value function loss \
+                    as the auxiliary objective, thereby sharing features between the policy and value function\
+                    while minimizing distortions to the policy
+                - aux_value_loss (:obj:`float`): The auxiliary value loss, we need to train the value network extra \
+                    during the auxiliary phase, it's the value loss we train the value network during auxiliary phase
+                - behavioral_cloning_loss (:obj:`float`): The behavioral cloning loss, used to optimize the auxiliary\
+                     objective while otherwise preserving the original policy
+        """
+        data = self._data_preprocess_learn(data)
+        # ====================
+        # PPG forward
+        # ====================
+        self._learn_model.train()
+        return_infos = []
+        if self._value_norm:
+            unnormalized_return = data['adv'] + data['value'] * self._running_mean_std.std
+            data['return'] = unnormalized_return / self._running_mean_std.std
+            self._running_mean_std.update(unnormalized_return.cpu().numpy())
+        else:
+            data['return'] = data['adv'] + data['value']
+
+        for epoch in range(self._cfg.learn.actor_epoch_per_collect):
+            for batch in split_data_generator(data, self._cfg.learn.batch_size, shuffle=True):
+                policy_data = batch
+                policy_adv = policy_data['adv']
+                if self._adv_norm:
+                    # Normalize advantage in a total train_batch
+                    policy_adv = (policy_adv - policy_adv.mean()) / (policy_adv.std() + 1e-8)
+                # Policy Phase(Policy)
+                policy_output = self._learn_model.forward(policy_data['obs'], mode='compute_actor')
+                policy_error_data = ppo_policy_data(
+                    policy_output['logit'], policy_data['logit'], policy_data['action'], policy_adv,
+                    policy_data['weight']
+                )
+                ppo_policy_loss, ppo_info = ppo_policy_error(policy_error_data, self._clip_ratio)
+                policy_loss = ppo_policy_loss.policy_loss - self._entropy_weight * ppo_policy_loss.entropy_loss
+                self._optimizer_ac.zero_grad()
+                policy_loss.backward()
+                self._optimizer_ac.step()
+
+        for epoch in range(self._cfg.learn.critic_epoch_per_collect):
+            for batch in split_data_generator(data, self._cfg.learn.batch_size, shuffle=True):
+                value_data = batch
```

Comment:
remove this line, just use `for value_data in split_data_generator(data, self._cfg.learn.batch_size, shuffle=True):`